src/lib/Parser.hs

```diff
@@ -341,34 +359,41 @@ interfaceInstance = do
   (p, pos) <- withPos letPat
   ann <- annot uType
   keyWord WhereKW
-  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
-  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
-  return $ ULet InstanceLet (p, ann') rhs'
+  rhsEither <- mkConstructorCall ann <$> (withSrc $ (URecord . NoExt) <$>
+      interfaceRecordFields "=")
+  case rhsEither of
+    Left err  -> fail err
+    Right rhs -> let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+                 in return $ ULet InstanceLet (p, ann') rhs'
   where
     -- Here, we are traversing the type annotation to retrieve the name of
     -- the interface and generate its corresponding constructor.
-    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
-    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
+    mkConstructorCall (WithSrc _ (UPi _ arr typ)) record
+      | arr `elem` [ClassArrow, ImplicitArrow] = mkConstructorCall typ record
+    -- The innermost UApp is an application of a variable that has the name of
+    -- the interface to a type. We retrieve its name and derive the name of
+    -- its automatically generated constructor, which we apply to the record
+    -- passed as a parameter.
+    mkConstructorCall (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) record =
+      Right $ (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` record
+    -- Given the instance "I Int Float" of a multi-parameter interface I, we
+    -- end up with something similar to UApp _ (UApp _ I Int) Float. The below
+    -- traverses UApps to get to the innermost one..
+    mkConstructorCall (WithSrc _ (UApp _ func _)) record =
+      mkConstructorCall func record
+    -- In this case, the type annotation does not contain any function
+    -- application; this indicates a Kind error.
+    mkConstructorCall _ _ =
+      Left ("Interface constructor was not applied to any type and can not " ++
+            "be instantiated")
```

Comment:
I'm wondering if there's a more direct algorithm here that would be less sensitive to edge cases. It seems like the logic we want is:
- First remove any class and implicit arrows
- Next, remove any applied arguments
- If we are left with a single identifier, that must be the constructor name; otherwise, the user has given us something we can't use, so throw an error saying that the type of the instance must be an interface constructor literal (applied to arguments with possible implicit arrows).

Perhaps this function could be written to do that directly, instead of just repeatedly pattern-matching as it does now?